src/psyclone/psyir/backend/base.py

```diff
@@ -174,3 +223,86 @@ def _visit(self, node):
             raise VisitorError(
                 "Unsupported node '{0}' found: method names attempted were "
                 "{1}.".format(type(node).__name__, str(possible_method_names)))
+
+    def loop_node(self, node):
+        '''This method is called when a Loop instance is found in the
+        PSyIR tree.
+
+        :param node: a Loop PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.Loop`
+
+        :returns: the loop node converted into a (language specific) string.
+        :rtype: str
+
+        '''
+        start = self._visit(node.start_expr)
+        stop = self._visit(node.stop_expr)
+        step = self._visit(node.step_expr)
+        variable_name = node.variable_name
+
+        self._depth += 1
+        body = ""
+        for child in node.loop_body:
+            body += self._visit(child)
+        self._depth -= 1
+
+        return self.do_loop_format.format(self._nindent, variable_name,
+                                          start, stop, step, body)
+
+    @abc.abstractproperty
+    def do_loop_format(self):
+        '''Returns the format for a do/for loop. The format variables
+        will be replaced as follows:
+        0: indentation string
+        1: loop variable
+        2: first loop iteration
+        3: last loop iteration
+        4: step size
+        5: body of the loop
+        This is used by base.loop_node() to create a loop representation.
+
+        :return: the format of a loop.
+        :rtype: str
+        '''
+
+    @abc.abstractproperty
+    def directive_start(self):
+        '''Returns the start of a directive, e.g. "#pragma ...\n{" in C,
+        or "!$" in Fortran. The string {0} in the result will be replaced
+        with the actual directive. It must be implemented by any visitor.
+
+        :return: the start of a directive.
+        :rtype: str
+        '''
+
+    @abc.abstractproperty
+    def directive_end(self):
```

Comment:
Again, I don't think a pure, abstract handler wins us anything (it will just change the error produced by the visitor pattern).

Reply:
The directive_start/_end methods are not called using the visitor patterns, they are not handlers, but mainly helper functions, called from base. So I think there would be some benefit to of getting a pylint warning if those are not implemented. Though of course if we would envision a backend that does not support openmp directives, then it would indeed not make sense to enforce having this function everywhere. But I don't care one way or another.

Reply:
Oh yes, I missed that they weren't handlers - that's OK then.